src/drivers/amd/vangogh/acp_sp_dma.c

```diff
@@ -0,0 +1,300 @@
+		/* SP Transmit FIFO Address and FIFO Size*/
+		sp_fifo_addr = (uint32_t)(&pscratch_mem_cfg->acp_transmit_fifo_buffer);
+		sp_fifo_addr = (sp_fifo_addr & ACP_DRAM_ADDRESS_MASK);
+		sp_fifo_addr = (sp_fifo_addr - ACP_DRAM_PHY_TRNS);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_FIFOADDR), sp_fifo_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_FIFOSIZE), (uint32_t)(SP_FIFO_SIZE));
+
+		/* Transmit RINGBUFFER Address and size*/
+		config->elem_array.elems[0].src =
+			(config->elem_array.elems[0].src & ACP_DRAM_ADDRESS_MASK);
+		sp_buff_addr = (config->elem_array.elems[0].src | ACP_DRAM_ADDR_TRNS);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_RINGBUFADDR), sp_buff_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_RINGBUFSIZE), sp_buff_size);
+
+		/* Transmit DMA transfer size in bytes */
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_DMA_SIZE),
+			     (uint32_t)(ACP_DMA_TRANS_SIZE_128));
+
+		/* Watermark size for SP transmit FIFO - Half of SP buffer size */
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_TX_INTR_WATERMARK_SIZE),
+			     (sp_buff_size >> 1));
+
+	} else if (config->direction == DMA_DIR_DEV_TO_MEM) {
+		/* SP Receive FIFO Address and FIFO Size*/
+		sp_fifo_addr = (uint32_t)(&pscratch_mem_cfg->acp_receive_fifo_buffer);
+		sp_fifo_addr = (sp_fifo_addr & ACP_DRAM_ADDRESS_MASK);
+		sp_fifo_addr = (sp_fifo_addr - ACP_DRAM_PHY_TRNS);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_FIFOADDR), sp_fifo_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_FIFOSIZE),
+			     (uint32_t)(SP_FIFO_SIZE));
+
+		/* Receive RINGBUFFER Address and size*/
+		config->elem_array.elems[0].dest =
+			(config->elem_array.elems[0].dest & ACP_DRAM_ADDRESS_MASK);
+		sp_buff_addr = (config->elem_array.elems[0].dest | ACP_DRAM_ADDR_TRNS);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_RINGBUFADDR), sp_buff_addr);
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_RINGBUFSIZE), sp_buff_size);
+
+		/* Receive DMA transfer size in bytes */
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_DMA_SIZE),
+			     (uint32_t)(ACP_DMA_TRANS_SIZE_128));
+
+		/* Watermark size for SP receive fifo - Half of SP buffer size*/
+		io_reg_write((PU_REGISTER_BASE + ACP_I2S_RX_INTR_WATERMARK_SIZE),
+			     (sp_buff_size >> 1));
+
+	} else {
+		tr_err(&acp_sp_vgh_tr, "DMA Config channel direction undefined %d",
+		       channel->direction);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int acp_dai_sp_dma_get_data_size(struct dma_chan_data *channel,
+				 uint32_t *avail, uint32_t *free)
+{
+	uint64_t tx_low, curr_tx_pos, tx_high;
+	uint64_t rx_low, curr_rx_pos, rx_high;
+
+	if (channel->direction == DMA_DIR_MEM_TO_DEV) {
+		tx_low = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_I2S_TX_LINEARPOSITIONCNTR_LOW);
+		tx_high = (uint32_t)io_reg_read(PU_REGISTER_BASE +
+				ACP_I2S_TX_LINEARPOSITIONCNTR_HIGH);
+		curr_tx_pos = (uint64_t)((tx_high << 32) | tx_low);
+		prev_tx_pos = curr_tx_pos;
+		*free = (sp_buff_size >> 1);
+		*avail = (sp_buff_size >> 1);
```

Comment:
this is strange - why all those register reads and calculations if you just return half-full? Same below

Reply:
These calculations are required for timer-based scheduling. we are in a process of enabling timer-based scheduling.

Reply:
> we are in a process of enabling timer-based scheduling.

Add a "TODO" or "Work in progress" comment in the code. Readers should never wonder why some code has no effect. They're more likely to guess it's old and dead code.

Reply:
Done